### ISSUE TITLE
`reuse_compilation_vms: true` breaks `workers` limit in `compilation`

### ISSUE THREAD

**Issue by voelzmo** [cloudfoundry/bosh#1108]
Steps to reproduce:
* specify a `compilation` section in your deployment manifest like this
```
compilation:
  workers: 2
  network: default
  reuse_compilation_vms: true
  cloud_properties:
    instance_type: c1.medium
    availability_zone: us-east-1c
```
* deploy something with many packages to compile, e.g. Cloud Foundry

expected:
* at most 2 compilation VMs are active at any given time

actual:
* More compilation VMs are created

Note:
* If `reuse_compilation_vms` is set to `false` everything works as expected.

**Comment by cunnie** [comment]
@voelzmo , We were unable to reproduce your problem. Perhaps you could send us a redacted manifest  and the version of BOSH Director (e.g. release 250) you were using? We were using release 250.

We had the following stanza in our test release:
```
compilation:
  workers: 2
  network: cf_private
  reuse_compilation_vms: true
  cloud_properties:
    availability_zone: us-east-1c
    instance_type: c3.large
```

And during our deploy (of CF) only 2 compilation VMs were spun up (expected behavior, not buggy behavior):

```
bosh -t xxxx vms
Acting as user 'admin' on 'xxxx'
Deployment `cf'

Director task 8

Task 8 done

+-------------------------------------------------------------------------------------------+---------+-----+---------+--------------+
| VM                                                                                        | State   | AZ  | VM Type | IPs          |
+-------------------------------------------------------------------------------------------+---------+-----+---------+--------------+
| compilation-656d3bb9-2b4e-4513-bb26-e1f78e82918d/0 (e216d84e-47bf-424a-87da-5faf578e3b4f) | running | n/a |         | 10.10.130.11 |
| compilation-d22dc122-423d-4f1d-8f63-fac1f1bf229f/0 (1d4ae658-0f16-49b8-b1d3-9f6c99052f1f) | running | n/a |         | 10.10.130.10 |
+-------------------------------------------------------------------------------------------+---------+-----+---------+--------------+

VMs total: 2
```

We created an integration test to specifically expose this type of problem, although we only committed to a branch not to *develop* or *master*: 
https://github.com/cloudfoundry/bosh/tree/compilation-112544589

**Comment by cunnie** [comment]
Hi @voelzmo ,

Tyler said that this *may* be related to this tracker story; the problem was fixed in release 248/build 3181:

https://www.pivotaltracker.com/story/show/111108272

—@yuzhangcmu & Brian Cunnie

**Comment by voelzmo** [comment]
@cunnie @yuzhangcmu Seems like this is fixed in the newer Bosh version. Tried to reproduce it with v250, but couldn't. So let's close this one, I'll let you know if I encounter it again. Thanks for investigating!

**Comment by tylerschultz** [comment]
We ended up figuring out how to recreate this issue. There are a couple of cases that cause the problem. One problem was that resource pool stemcell references were considered different, even if they refer to the same stemcell. The other problem was that the pooling logic was bad and would not destroy instances from other stemcells when the pool was full.

**Comment by voelzmo** [comment]
Awesome, happy you guys poked at it long enough to find what caused this. Thanks!

**Comment by datianshi** [comment]
Is this closed? I am using the bosh release 250 with a bosh deployment on azure. Hit this exact issue

**Comment by tylerschultz** [comment]
I believe the fix was in release v255.1.
